### ISSUE TITLE
Missing bitsplat pattern optimization

### ISSUE THREAD

**Issue by davidbolvansky** [llvm/llvm-project#53610]
```

typedef int v4i32 __attribute__((vector_size(16)));
#define bit 3

int get_cmpmasks(int mask)
{
    mask >>= bit;
    mask &= 1;
    mask = -mask;
    return mask;
}
```

LLVM:
get_cmpmasks(int):                      # @get_cmpmasks(int)
        mov     eax, edi
        shr     eax, 3
        and     eax, 1
        neg     eax
        ret

GCC:
get_cmpmasks(int):
        mov     eax, edi
        sal     eax, 28
        sar     eax, 31
        ret

https://godbolt.org/z/n71YnhKa9

**Comment by RKSimon** [comment]
General Proof: https://alive2.llvm.org/ce/z/ZcYiqe

**Comment by RKSimon** [comment]
@rotateright @LebedevRI Should this be handled in the negator?

**Comment by rotateright** [comment]
Just to be clear, 'nsw' isn't necessary:
https://alive2.llvm.org/ce/z/T8fqyw

Not sure if a general subtraction would be better in IR/analysis than a negate in that pattern, so we might make this a shift-by-constant-amount fold only. That also gives us more flexibility if there's an extra use of the 'lshr'. We can still do the transform since we're reducing instructions in the base case.

**Comment by RKSimon** [comment]
Yeah, the non-constant case is probably too much - I started playing with non-uniform (constant) vector shift codegen and sort of didn't stop......

**Comment by RKSimon** [comment]
Candidate Patch: https://reviews.llvm.org/D119715

**Comment by RKSimon** [comment]
Fixed by 9606c690873fab93d895b2b763d8e2ed781c05af

**Comment by LebedevRI** [comment]
Thanks!
